Google KMS integration

roysagi opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16136

## Description
Google KMS integration.

Currently missing:
 * Integration Image

## Does it break backward compatibility?
   - No

## Must have
- [x] Tests
- [ ] Documentation (with link to it)
- [x] Code Review

## Additional changes
Created a new docker: https://github.com/demisto/dockerfiles/pull/535
holds the required packages to use in the integration

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml)
- [ ] [CHANGELOG](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/CHANGELOG.md)
- [ ] [Description](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/GoogleKeyManagementService_description.md)

roysagi commented:
@ronykoz @David-BMS - FYI

yaakovi commented on Integrations/GoogleKeyManagementService/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+##[Unreleased]  
```
remove this file

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+    def _init_kms_client(self):
+        """Creates the Python API client for Google Cloud KMS using service account credentials.
+        """
+        cur_directory_path = os.getcwd()
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+    Returns:
+        Dict from string representation.
+    """
+    if string is None or len(string) == 0:
```
changed

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+        key, value = section.split(':')
+        arg_dict[key] = value
+
+    return arg_dict
```
This does not work... I added the split(':',1) correction though.

yaakovi replied:
what do you mean "doesn't work"?
what is the problem with it?
![image](https://user-images.githubusercontent.com/30797606/67264212-1c24ae00-f4b3-11e9-8c66-4d69ddc6f2a4.png)

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+    Returns:
+        Dict to be used as body for update command
+    """
+    body = {}  # type:Dict[str,Any]
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+        Dict to be used as body for update command
+    """
+    body = {}  # type:Dict[str,Any]
+    for field in update_mask:
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+    body = {}  # type:Dict[str,Any]
+    for field in update_mask:
+        if '.' in field:
+            split_field = field.split('.')
```
the function change removed these splits

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+    }
+
+
+def init_dict(body: Dict[str, Any], parent: str, key: str, value: Any):
```
removed function

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+
+    # Algorithm must be a "DECRYPT" type asymmetric algorithm - if not, raise an error to the user.
+    if 'DECRYPT' not in key_algo:
+        raise Exception(f"{crypto_key_version_name} is not a valid asymmetric CryptoKeyVersion")
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+        if command == 'test-module':
+            test_function(client)
+            demisto.results('ok')
+            sys.exit(0)
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+                                                                                           'Encrypter/Decrypter',
+                                                                                           'Project-Admin']:
+            str_results = symmetric_encrypt_key_command(client, demisto.args())
+            demisto.results(f"The text has been encrypted.\nCiphertext: {str_results}")
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,882 @@
+            demisto.results('ok')
+            sys.exit(0)
+
+        elif command == f'{INTEGRATION_COMMAND_NAME}create-key' and client.role in ['Project-Admin', 'KMS-Admin']:
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml:
```diff
@@ -0,0 +1,960 @@
+  name: proxy
+  required: false
+  type: 8
+description: Integration Template
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml:
```diff
@@ -0,0 +1,960 @@
+      description: The current state of the CryptoKeyVersion.
+      type: String
+    - contextPath: GoogleKMS.CryptoKey.PrimaryCryptoKeyVersion.CreationTime
+      description: ' The time at which this CryptoKeyVersion was created.'
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml:
```diff
@@ -0,0 +1,960 @@
+
+        A base64-encoded string.
+      isArray: false
+      name: "additional_authenticated_data\t "
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml:
```diff
@@ -0,0 +1,960 @@
+      description: The current state of the CryptoKeyVersion.
+      type: String
+    - contextPath: GoogleKMS.CryptoKey.PrimaryCryptoKeyVersion.CreationTime
+      description: ' The time at which this CryptoKeyVersion was created.'
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml:
```diff
@@ -0,0 +1,960 @@
+      description: The current state of the CryptoKeyVersion.
+      type: String
+    - contextPath: GoogleKMS.CryptoKey.PrimaryCryptoKeyVersion.CreationTime
+      description: ' The time at which this CryptoKeyVersion was created.'
```
done

roysagi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService_test.py:
```diff
@@ -0,0 +1,69 @@
+    assert created_dict['label2'] == 'value2'
+
+
+def test_ags_extract():
```
done

roysagi commented on TestPlaybooks/playbook-Google-KMS-test.yml:
```diff
@@ -0,0 +1,648 @@
+      name: google-kms-get-key
+      description: Returns metadata for a given CryptoKey, as well as its primary
+        CryptoKeyVersion.
+      script: Google Key Management Service|||google-kms-get-key
```
done

roysagi commented on TestPlaybooks/playbook-Google-KMS-test.yml:
```diff
@@ -0,0 +1,648 @@
+      name: google-kms-get-key
+      description: Returns metadata for a given CryptoKey, as well as its primary
+        CryptoKeyVersion.
+      script: Google Key Management Service|||google-kms-get-key
```
done

roysagi commented on TestPlaybooks/playbook-Google-KMS-test.yml:
```diff
@@ -0,0 +1,648 @@
+      name: google-kms-get-key
+      description: Returns metadata for a given CryptoKey, as well as its primary
+        CryptoKeyVersion.
+      script: Google Key Management Service|||google-kms-get-key
```
done

roysagi commented on TestPlaybooks/playbook-Google-KMS-test.yml:
```diff
@@ -0,0 +1,648 @@
+      name: google-kms-get-key
+      description: Returns metadata for a given CryptoKey, as well as its primary
+        CryptoKeyVersion.
+      script: Google Key Management Service|||google-kms-get-key
```
done

roysagi commented on TestPlaybooks/playbook-Google-KMS-test.yml:
```diff
@@ -0,0 +1,648 @@
+      version: -1
+      name: google-kms-list-keys
+      description: Lists all keys in key ring
+      script: Google Key Management Service|||google-kms-list-keys
```
done

roysagi commented:
nevermind. does work. my bad

On Tue, Oct 22, 2019 at 10:04 AM Shai Yaakovi <notifications@github.com>
wrote:

> *@yaakovi* commented on this pull request.
> ------------------------------
>
> In Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_demisto_content_pull_4500-23discussion-5Fr337354139&d=DwMCaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=7ewWo63KHbJvO0-WWrKDEtIECENX6iTJQQS3hs10q5Y&m=vVMmmR1Y0fLAD6iVdbiOkYpUfi2t6aA5sGYC8pZWjVc&s=T4W0iNCUdSvP2NMMPlP7dCj8ncFBw4L2yMNUshq9Tu8&e=>
> :
>
> > +    split_string = string.split(',')
> +    arg_dict = {}
> +    for section in split_string:
> +        key, value = section.split(':')
> +        arg_dict[key] = value
> +
> +    return arg_dict
>
> what do you mean "doesn't work"?
> what is the problem with it?
> [image: image]
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__user-2Dimages.githubusercontent.com_30797606_67264212-2D1c24ae00-2Df4b3-2D11e9-2D8c66-2D4d69ddc6f2a4.png&d=DwMCaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=7ewWo63KHbJvO0-WWrKDEtIECENX6iTJQQS3hs10q5Y&m=vVMmmR1Y0fLAD6iVdbiOkYpUfi2t6aA5sGYC8pZWjVc&s=Ag_-oL2Z9ZH3W5-L2FSZOSezjNiSJxOmF9wIu7rus4o&e=>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_demisto_content_pull_4500-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAL7XIEXAHPQQ6JM2M4QT5NLQP2QXNA5CNFSM4I23WRP2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIW5OJA-23discussion-5Fr337354139&d=DwMCaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=7ewWo63KHbJvO0-WWrKDEtIECENX6iTJQQS3hs10q5Y&m=vVMmmR1Y0fLAD6iVdbiOkYpUfi2t6aA5sGYC8pZWjVc&s=3Gmbl5ZmU8R3SQxbgV0FJZwFPqtedalVSMLM53zS574&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AL7XIEXP2RKPDE7CQ55THN3QP2QXNANCNFSM4I23WRPQ&d=DwMCaQ&c=V9IgWpI5PvzTw83UyHGVSoW3Uc1MFWe5J8PTfkrzVSo&r=7ewWo63KHbJvO0-WWrKDEtIECENX6iTJQQS3hs10q5Y&m=vVMmmR1Y0fLAD6iVdbiOkYpUfi2t6aA5sGYC8pZWjVc&s=bmgzVe8NTXWa99sWTTp7bZPBU7h7xZcH-_IHJ_rq2mM&e=>
> .
>

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+            raise Exception(f"Your Service Account Role does not permit the use of {command} command.")
+
+        results = cmd_func(client, demisto.args())  # type: ignore
+        if results:
```
you should always return results (otherwise the command will never finish

roysagi replied:
done

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+        cmd_role = list(COMMANDS.get(command)[1])
+
+        if client.role not in cmd_role:
+            raise Exception(f"Your Service Account Role does not permit the use of {command} command.")
```
```suggestion
            raise PermissionError(f"Your Service Account Role does not permit the use of {command} command.")
```

roysagi replied:
done

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+        response = client.kms_client.list_key_rings(location_path)
+
+        # Count the number of existing KeyRings.
+        if sum(1 for ring in response) == 0:
```
will those not work?
`if not response:`
`if not list(response):`

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+
+    # Print results
+    response = client.kms_client.update_crypto_key_version(version, update_mask)
+    demisto.results(f'CryptoKeyVersion {crypto_key_version_name}\'s state has been set to '
```
this should be a `return`, not a `demisto.results`.
same for everywhere.
use `return (<string>, )` or `return (<string>, None, None)`

roysagi replied:
done

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+        key, value = section.split(':', 1)
+        arg_dict[key] = value
+
+    return dict()
```
i'm guessing this is a bug

roysagi replied:
yeah... a revert issue - fixed

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,854 @@
+    return str(project_id), str(location_id), str(key_ring_id), str(crypto_key_id)
+
+
+def get_update_mask(args: Dict[str, Any]) -> Dict:
```
still not sure why you need this function... let's discuss in person

roysagi replied:
This is part of the API's update request - look here for examples: https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/kms/api-client/snippets.py

roysagi commented:
Hey @kirbles19 if you can please take a look at these files:
- [ ] [YAML file](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/GoogleKeyManagementService.yml)
- [ ] [CHANGELOG](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/CHANGELOG.md)
- [ ] [Description](https://github.com/demisto/content/blob/google_kms/Integrations/GoogleKeyManagementService/GoogleKeyManagementService_description.md)

@yaakovi @ronykoz - FYI

ronykoz commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,956 @@
+
+def main():
+    COMMANDS = {
+        f'{INTEGRATION_COMMAND_NAME}create-key': (create_crypto_key_command, ['Project-Admin', 'KMS-Admin']),
```
It will look nicer to store it this way, and update the CONST accrodingly
```suggestion
        f'{INTEGRATION_COMMAND_NAME}-create-key': (create_crypto_key_command, ['Project-Admin', 'KMS-Admin']),
```

roysagi replied:
will do

ronykoz commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -0,0 +1,956 @@
+
+        results = cmd_func(client, demisto.args())  # type: ignore
+        return_outputs(*results)
+        return_error("this should kill the playbook")
```
what?

ronykoz commented:
@yaakovi can you go over it one last time?

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -81,28 +82,80 @@ def arg_dict_creator(string: str):
+        return None
+
+    cleared_labels = {}  # type:Dict
+    for label in labels.keys():
```
`for label in labels:`

can you give an example of the original labels value?
why the magic numbers? 1, 2, -1

roysagi replied:
will address that

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -404,11 +486,21 @@ def symmetric_encrypt_key_command(client: Client, args: Dict[str, Any]) -> Tuple
+    elif args.get('entry_id'):
+        file = demisto.getFilePath(args.get('entry_id'))
+        file_path = file['path']
+        fp = open(file_path, 'rb')
```
use `with`

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -438,7 +545,22 @@ def symmetric_decrypt_key_command(client: Client, args: Dict[str, Any]) -> Tuple
+    elif args.get('entry_id'):
+        file = demisto.getFilePath(args.get('entry_id'))
+        file_path = file['path']
+        fp = open(file_path, 'rb')
```
use `with`

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -695,11 +841,21 @@ def asymmetric_encrypt_command(client: Client, args: Dict[str, Any]) -> Tuple[st
+    elif args.get('entry_id'):
+        file = demisto.getFilePath(args.get('entry_id'))
+        file_path = file['path']
+        fp = open(file_path, 'rb')
```
use `with`

yaakovi commented on Integrations/GoogleKeyManagementService/GoogleKeyManagementService.py:
```diff
@@ -754,73 +925,231 @@ def asymmetric_decrypt_command(client: Client, args: Dict[str, Any]) -> Tuple[st
+    elif args.get('entry_id'):
+        file = demisto.getFilePath(args.get('entry_id'))
+        file_path = file['path']
+        fp = open(file_path, 'rb')
```
use `with`

richardbluestone commented:
I've reviewed and updated the yml file

richardbluestone commented:
Updated mk and yml files. You can now merge.
